### PR TITLE
Document breaking change rule for the underlying type of an enum

### PR DESCRIPTION
**Pull request by jnm2**
Following up from https://github.com/dotnet/corefx/issues/31889#issuecomment-418461790.

/cc @nguerrera, @weshaggard, @joperezr

### CODE REVIEW

**Comment by svick** [comment]
I'm pretty sure this should be in the "Disallowed" section, not in "Allowed". :slightly_smiling_face:

Also, this new rule pretty much implies (by omission) that changing the underlying type of an enum is allowed, if it doesn't change size (e.g. changing from `int` to `uint`). Is that really the case?

**Comment by jnm2** [comment]
@svick Lol, thanks!

> Also, this new rule pretty much implies (by omission) that changing the underlying type of an enum is allowed, if it doesn't change size (e.g. changing from int to uint). Is that really the case?

I thought about this and couldn't come up with a reason to disallow switching between same-sized underlying types, so I hoped you or someone would know.

**Comment by jnm2** [comment]
Thought prompted by @jp2masa: switching between signed and signed versions of an underlying type would be redefining the numeric value of any fields with the high bit set. (Does that numeric value matter for flags enums?)

**Comment by AtsushiKan** [comment]
I believe this rule:

> Changing the type of a property, field, parameter or return value

already makes changing underlying enum types disallowed. 


It's worth calling out separately since most people probably don't deconstruct enums this way but I don't think we to split out different-size vs same size or justify the rule with custom attribute blobs.

**Comment by jnm2** [comment]
@AtsushiKan I'd like to call out that in the case of enum fields, assemblies become partially unparseable. That's a unique kind of problem among breaking changes.

**Comment by svick** [comment]
@jnm2 

> I thought about this and couldn't come up with a reason to disallow switching between same-sized underlying types, so I hoped you or someone would know.

I think it's a source-breaking change. If you have `enum IntEnum : int {}` and `enum UIntEnum : uint {}`, then `(IntEnum)(-1)` compiles, but `(UIntEnum)(-1)` doesn't. This means that changing `enum E` from `int` to `uint` would stop `(E)(-1)` from compiling.

**Comment by jkotas** [comment]
> I think it's a source-breaking change. 

It is behavior change too. Consider `int x = (int)Enum.ToObject(typeof(MyEnum), 0)`. It works when MyEnum is int. It breaks if MyEnum underlying type is changed to uint.

File: Documentation/coding-guidelines/breaking-change-rules.md [hunk 1]
```diff
@@ -163,6 +163,12 @@ Breaking Change Rules
 
 * Changing a `struct` type to a `ref struct` type and vice versa
 
+* Changing the underlying type of an enum to a different size
+
+    This is a binary breaking change in a category all its own: not just at runtime but also while passively parsing the assembly’s attributes.
+
+    The ECMA 335 format requires the reader to resolve the enum’s assembly and look up the enum’s underlying data size, given the enum’s serialized type name, in order to know how many bytes to read as the attribute value. So if a value of the enum has been used as an attribute argument, and the enum data size in the resolved assembly differs from the enum data size when the assembly was compiled, the assembly reader will have no option but to read too much or too little and misread the current attribute argument and all following arguments. (The reader may not necessarily realize it is misinterpreting the data unless the resolved enum data size is smaller, which causes the reader to read past the end of the available data.)
```
Comment (jkotas) [inline Documentation/coding-guidelines/breaking-change-rules.md]: There are many more main stream cases that this breaks. I do not think that highlighting the custom attribute case here helps clarity.

**Comment by jnm2** [comment]
Thanks for the reviews! How does this look?

**Comment by weshaggard** [comment]
Thanks @jnm2

**Pull request merged**